### PR TITLE
chore(gevent): deprecate gevent context provider

### PR DESCRIPTION
**Pull request by mabdinur**
This PR deprecates `ddtrace.contrib.gevent.provider.GeventContextProvider` in favor of the default context provider.

`greenlet>=1` and `gevent>=1.5` introduced support for contextvars. With this change we no longer need to maintain out own gevent aware contextvariables. 

Note: By default contextvars are isolated to individual greenlets. When a greenlet is spawned, the context of the parent greenlet is not shared. This ensures contextvars act as "greenlet-local" storage. This behavior breaks the ddtrace gevent integration. To support propagating trace information to spawned greenlets the `ddtrace` gevent integration uses a `TracingMixin` to copy/propagate ddtrace trace context. 

With this change the minimum supported gevent version is [v20.12.0](https://www.gevent.org/changelog.html#id36)

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [x] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/contrib/gevent/patch.py [hunk 1]
```diff
@@ -34,8 +31,8 @@ def patch():
         return
     gevent.__datadog_patch = True
 
+    ddtrace.Pin().onto(gevent)
```
Comment (mabdinur) [inline ddtrace/contrib/gevent/patch.py:34]:
```suggestion
```

File: docs/index.rst [hunk 2]
```diff
@@ -90,7 +90,7 @@ contacting support.
 +--------------------------------------------------+---------------+----------------+
 | :ref:`flask_login`                               | >= 0.6.2      | Yes [6]_       |
 +--------------------------------------------------+---------------+----------------+
-| :ref:`gevent`                                    | >= 1.5        | No             |
+| :ref:`gevent` (greenlet>=1.0)                    | >= 1.5        | No             |
```
Comment (mabdinur) [inline docs/index.rst:93]:
```suggestion
| :ref:`gevent` (greenlet>=1.0)                    | >= 20.9        | No             |
```

File: tests/contrib/gevent/test_tracer.py [hunk 3]
```diff
@@ -76,7 +76,7 @@ def greenlet(_):
                 gevent.sleep(0.01)
 
         funcs = [
-            gevent.pool.Group().map,
+            gevent.pool.Group().imap_unordered,
             gevent.pool.Group().imap,
             gevent.pool.Group().imap_unordered,
```
Comment (erikayasuda) [inline tests/contrib/gevent/test_tracer.py:81]: Why do we have this one twice?

File: riotfile.py [hunk 4]
```diff
@@ -2362,31 +2360,24 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
-                                "gevent": "~=21.8.0",
-                            },
+                            pkgs={"gevent": "~=21.8.0"},
                         ),
                         Venv(
                             pys="3.11",
-                            pkgs={
-                                "gevent": "~=22.8.0",
-                            },
+                            pkgs={"gevent": "~=22.8.0"},
```
Comment (erikayasuda) [inline riotfile.py:2380]: Do these all have different versions of `gevent` just to have as much test coverage of 20.12.1+?

**Review by erikayasuda — COMMENTED** [review]
Some QQs

File: tests/contrib/gevent/test_tracer.py (hunk 3 above)
Comment (mabdinur) [reply tests/contrib/gevent/test_tracer.py:81]: I messed up. I will revert this line

File: tests/contrib/gevent/test_tracer.py [hunk 5]
```diff
@@ -76,7 +76,7 @@ def greenlet(_):
                 gevent.sleep(0.01)
 
         funcs = [
-            gevent.pool.Group().map,
+            gevent.pool.Group().imap_unordered,
```
Comment (mabdinur) [inline tests/contrib/gevent/test_tracer.py:79]:
```suggestion
            gevent.pool.Group().map,
```

File: riotfile.py (hunk 4 above)
Comment (mabdinur) [reply riotfile.py:2380]: These riot venvs test otel+opentracing compatibility. We will remove opentracing support in ddtrace v3.0 so I don't want to invest too much time into cleaning this up.

File: releasenotes/notes/drop-gevent-support-3da5ffc8e94f73ea.yaml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+---
+deprecations:
+  - |
+    gevent: Deprecates ``ddtrace.contrib.gevent.provider.GeventContextProvider``. Drops support for `gevent<20.12.0` and `greenlet<1.0`.
```
Comment (mabdinur) [inline releasenotes/notes/drop-gevent-support-3da5ffc8e94f73ea.yaml:4]:
```suggestion
    gevent: deprecates ``ddtrace.contrib.gevent.provider.GeventContextProvider``. Drops support for `gevent<20.12.0` and `greenlet<1.0`.
```

**Review by erikayasuda — APPROVED** [review]

File: docs/index.rst [hunk 7]
```diff
@@ -90,7 +90,7 @@ contacting support.
 +--------------------------------------------------+---------------+----------------+
 | :ref:`flask_login`                               | >= 0.6.2      | Yes [6]_       |
 +--------------------------------------------------+---------------+----------------+
-| :ref:`gevent`                                    | >= 1.5        | No             |
+| :ref:`gevent` (greenlet>=1.0)                    | >= 20.12      | No             |
```
Comment (emmettbutler) [inline docs/index.rst:93]: Changing the [corporate documentation](https://github.com/DataDog/documentation/blob/5894a774b43284706c10b5dcfcf248d1352eb527/content/en/tracing/trace_collection/compatibility/python.md) to match would reduce the possibility for confusion.

**Review by emmettbutler — APPROVED** [review]

File: docs/index.rst (hunk 7 above)
Comment (mabdinur) [reply docs/index.rst:93]: https://github.com/DataDog/documentation/pull/22033

File: ddtrace/contrib/gevent/greenlet.py [hunk 8]
```diff
@@ -1,33 +1,28 @@
...
 
 class TracingMixin(object):
     def __init__(self, *args, **kwargs):
-        # get the active context/span if available
-        current_g = gevent.getcurrent()
-        ctx = getattr(current_g, GeventContextProvider._CONTEXT_ATTR, None)
-
-        # create the Greenlet as usual
+        # Storse the current Datadog context.
+        # This is necessary to ensure tracing context is passed to greenlets.
```
Comment (mabdinur) [inline ddtrace/contrib/gevent/greenlet.py:12]:
```suggestion
        # This is necessary to ensure tracing context is passed to greenlets.
        # Avoids setting Greenlet.gr_context, setting field introduces unintended side-effects
        # in thirdparty library.
```

File: ddtrace/contrib/gevent/greenlet.py [hunk 9]
```diff
@@ -1,33 +1,30 @@
...
+        self.trace_context = _DD_CONTEXTVAR.get()
         super(TracingMixin, self).__init__(*args, **kwargs)
 
-        # copy the active span/context into the new greenlet
-        if ctx:
-            setattr(self, GeventContextProvider._CONTEXT_ATTR, ctx)
+    def run(self):
+        # Propagates Datadog context to spawned greenlets
+        _DD_CONTEXTVAR.set(self.trace_context)
+        super(TracingMixin, self).run()
```
Comment (r1viollet) [inline ddtrace/contrib/gevent/greenlet.py:21]: why do we not restore things to the parent context after run ?

Comment (r1viollet) [reply ddtrace/contrib/gevent/greenlet.py:21]: For context, this is not an issue right now, I was mostly curious if this made sense.

**Pull request merged**